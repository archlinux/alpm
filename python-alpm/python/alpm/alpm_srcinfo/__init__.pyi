"""A module for parsing and linting of ALPM SRCINFO files."""

from . import error, source_info
from .error import SourceInfoError
from .source_info.v1 import SourceInfoV1
from .source_info.v1.merged import MergedPackage

__all__ = [
    "SourceInfoError",
    "error",
    "source_info",
    "SourceInfoV1",
    "MergedPackage",
]
