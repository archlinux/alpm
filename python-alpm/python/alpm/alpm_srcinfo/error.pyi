"""Exceptions used by alpm_srcinfo module."""

class SourceInfoError(Exception):
    """The high-level exception that can occur when using alpm_srcinfo module."""

__all__ = [
    "SourceInfoError",
]
