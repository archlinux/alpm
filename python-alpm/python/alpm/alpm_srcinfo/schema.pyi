"""Schemas for SRCINFO data."""

from pathlib import Path
from typing import Union, TYPE_CHECKING

if TYPE_CHECKING:
    from alpm.alpm_types import SchemaVersion

class SourceInfoSchema:
    """SRCINFO schema.

    The schema of a SRCINFO refers to the minimum required sections and keywords, as
    well as the complete set of available keywords in a specific version.
    """

    def __init__(self, version: Union[str, "SchemaVersion"]) -> None:
        """Create a SourceInfoSchema from SchemaVersion or its string representation.

        Args:
            version (Union[str, SchemaVersion]): either a SchemaVersion or a string
                representation of SchemaVersion.

        Raises:
            ALPMError: if the string representation of the version is invalid.
            SourceInfoError: if there is no corresponding schema for the provided major
                version.

        """

    @staticmethod
    def derive_from_str(srcinfo: str) -> "SourceInfoSchema":
        """Derive the schema from a string containing SRCINFO data.

        Args:
            srcinfo (str): The srcinfo string to derive the schema from.

        Returns:
            SourceInfoSchema: The derived schema.

        Raises:
            SourceInfoError: if the srcinfo string cannot be parsed.

        """

    @staticmethod
    def derive_from_file(path: Union[str, Path]) -> "SourceInfoSchema":
        """Derive the schema from a file containing SRCINFO data.

        Args:
            path (Union[str, Path]): The path to the file containing SRCINFO data.

        Returns:
            SourceInfoSchema: The derived schema.

        Raises:
            SourceInfoError: if the file cannot be read or parsed.

        """

    @property
    def version(self) -> "SchemaVersion":
        """The schema version."""

    def __str__(self) -> str: ...
    def __repr__(self) -> str: ...
    def __eq__(self, other: object) -> bool: ...

__all__ = [
    """SourceInfoSchema""",
]
