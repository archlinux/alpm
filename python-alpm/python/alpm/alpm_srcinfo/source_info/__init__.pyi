"""Data representations and integrations for reading of SRCINFO data."""

from . import v1

__all__ = ["v1"]
