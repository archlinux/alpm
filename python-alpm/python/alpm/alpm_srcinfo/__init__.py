# TODO: <https://gitlab.archlinux.org/archlinux/alpm/alpm/-/issues/204>
from alpm import alpm_srcinfo as _mod

__all__ = _mod.__all__
for name in getattr(_mod, "__all__"):
    globals()[name] = getattr(_mod, name)
