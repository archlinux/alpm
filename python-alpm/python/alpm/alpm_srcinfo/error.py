# TODO: <https://gitlab.archlinux.org/archlinux/alpm/alpm/-/issues/204>
from alpm.alpm_srcinfo import error as _mod

__all__ = _mod.__all__
for name in getattr(_mod, "__all__"):
    globals()[name] = getattr(_mod, name)
