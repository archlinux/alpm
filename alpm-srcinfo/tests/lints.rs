use std::{fs::read_to_string, path::PathBuf};

use alpm_srcinfo::{Error, SourceInfo};
use insta::assert_snapshot;
use rstest::rstest;
use testresult::TestResult;

/// .SRCINFO files are processed in two steps:
///
/// - Raw parsing step with winnow into intermediate representation (IR)
/// - Bring IR into proper struct representation and apply lints
///
/// This test tests linting issues and error messages during the second step.
#[rstest]
pub fn ensure_lints(#[files("tests/lints/*")] case: PathBuf) -> TestResult {
    // Read the input file and parse it.

    let input = read_to_string(&case)?;
    let source_info_result = SourceInfo::from_string(input.as_str());

    // Make sure there're no parse errors
    let source_info_result = match source_info_result {
        Ok(result) => result,
        Err(err) => {
            return Err(format!(
                "The parser errored even though it should've succeeded the parsing step:\n{err}"
            )
            .into());
        }
    };

    // Ensure that there're lint errors
    let Some(lint_errors) = source_info_result.errors() else {
        return Err(
            "The parser didn't produce any lint errors, even though there should be some".into(),
        );
    };

    let error_msg = Error::SourceInfoErrors(lint_errors.clone()).to_string();

    let name = case.file_stem().unwrap().to_str().unwrap();

    // Run the tests with the input being displayed as the description.
    // This makes reviewing this whole stuff a lot easier.
    // Also remove the usual module prefix by explicitly setting the snapshot path.
    // This isn't necessary, as we're already manually sorting snapshots by test scenario.
    let input_clone = input.clone();
    insta::with_settings!({
        description => input_clone,
        snapshot_path => "lint_snapshots",
        prepend_module_to_snapshot => false,
    }, {
        assert_snapshot!(name, error_msg);
    });

    Ok(())
}
