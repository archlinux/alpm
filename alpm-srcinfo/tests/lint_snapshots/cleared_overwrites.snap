---
source: alpm-srcinfo/tests/lints.rs
description: "pkgbase = test\n    pkgver = 1.0.0\n\tpkgrel = 1\n    arch = any\n    source = https://domain.tld/testing/test.tar.gz\n    depends = glibc\n\npkgname = test\n    depends = llvm\n    # Depends clears a property that has just been set.\n    depends =\n\npkgname = test_other\n    depends =\n    # Any declaration in a package implicitly clears the default values,\n    # so this isn't necessary\n    depends = llvm\n"
expression: error_msg
---
Errors while parsing SRCINFO data:

[0] Linter Warning on line 9:
    9: |     depends = llvm

    This keyword is set and unset for this package. A keyword should either only be unset or overridden.

──────────────────────────────
[1] Linter Warning on line 17:
    17: |     depends = llvm

    This keyword is set and unset for this package. A keyword should either only be unset or overridden.
